### PR TITLE
Add long diagnostics for E0373

### PR DESCRIPTION
**Pull request by AlisdairO**
As title!

Part of #24407.
r? @Manishearth

### CODE REVIEW

File: src/librustc_borrowck/diagnostics.rs [hunk 1]
```diff
@@ -12,6 +12,56 @@
 
 register_long_diagnostics! {
 
+E0373: r##"
+This error occurs when an attempt is made to us data captured by a closure,
```
Comment (oli-obk) [inline src/librustc_borrowck/diagnostics.rs]: us -> use

File: src/librustc_borrowck/diagnostics.rs [hunk 2]
```diff
@@ -12,6 +12,56 @@
...
+return a closure:
+
+```
+fn foo() -> Box<Fn(u32) -> u32> {
+    let x = 0u32;
+    Box::new(|y| x + y)
+}
+```
+
+Notice that `x` is stack-allocated by `foo()`. By default, rust captures
```
Comment (apasel422) [inline src/librustc_borrowck/diagnostics.rs]: s/rust/Rust/

**Comment by Manishearth** [comment]
r=me with the nits above fixed

**Comment by AlisdairO** [comment]
nits fixed, squashed!

File: src/librustc_borrowck/diagnostics.rs [hunk 3]
```diff
@@ -12,6 +12,56 @@
 
 register_long_diagnostics! {
 
+E0373: r##"
+This error occurs when an attempt is made to use data captured by a closure,
```
Comment (nagisa) [inline src/librustc_borrowck/diagnostics.rs]: Nit: Its the attempt to capture which causes the error, not the use of the captured variable.

**Pull request merged**